client/db.go

```diff
@@ -0,0 +1,627 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Peter Mattis (peter.mattis@gmail.com)
+
+package client
+
+import (
+	"bytes"
+	"encoding"
+	"fmt"
+	"log"
+	"net/url"
+	"strconv"
+	"time"
+
+	"github.com/cockroachdb/cockroach/proto"
+	"github.com/cockroachdb/cockroach/testutils"
+	gogoproto "github.com/gogo/protobuf/proto"
+)
+
+// KeyValue ...
+type KeyValue struct {
+	key       []byte
+	value     interface{}
+	timestamp int64
+}
+
+func (kv *KeyValue) String() string {
+	switch t := kv.value.(type) {
+	case nil:
+		return string(kv.key) + "=nil"
+	case []byte:
+		return string(kv.key) + "=" + string(t)
+	case *int64:
+		return string(kv.key) + "=" + strconv.FormatInt(*t, 10)
+	}
+	return string(kv.key) + fmt.Sprintf("=<error:%T>", kv.value)
+}
+
+// Key ...
+func (kv *KeyValue) Key(k interface{}) error {
+	switch t := k.(type) {
+	case encoding.BinaryUnmarshaler:
+		return t.UnmarshalBinary(kv.key)
+	case *[]byte:
+		*t = kv.key
+	case *string:
+		*t = string(kv.key)
+	}
+	return fmt.Errorf("unable to unmarshal value: %T", k)
+}
+
+// KeyBytes ...
+func (kv *KeyValue) KeyBytes() []byte {
+	return kv.key
+}
+
+// Exists ...
+func (kv *KeyValue) Exists() bool {
+	return kv.value != nil
+}
+
+func (kv *KeyValue) setValue(v *proto.Value) {
+	if v == nil {
+		return
+	}
+	if v.Bytes != nil {
+		kv.value = v.Bytes
+	} else if v.Integer != nil {
+		kv.value = v.Integer
+	}
+	if ts := v.Timestamp; ts != nil {
+		kv.timestamp = ts.WallTime
+	}
+}
+
+// Value ...
+func (kv *KeyValue) Value(b encoding.BinaryUnmarshaler) error {
+	switch val := kv.value.(type) {
+	case []byte:
+		return b.UnmarshalBinary(val)
+	}
+	return fmt.Errorf("unable to unmarshal value: %T", kv.value)
+}
+
+// ValueBytes ...
+func (kv *KeyValue) ValueBytes() []byte {
+	return kv.value.([]byte)
+}
+
+// ValueInt ...
+func (kv *KeyValue) ValueInt() int64 {
+	return *kv.value.(*int64)
+}
+
+// ValueProto ...
+func (kv *KeyValue) ValueProto(msg gogoproto.Message) error {
+	switch val := kv.value.(type) {
+	case []byte:
+		return gogoproto.Unmarshal(val, msg)
+	}
+	return fmt.Errorf("unable to unmarshal proto: %T", kv.value)
+}
+
+// Timestamp ...
+func (kv *KeyValue) Timestamp() time.Time {
+	sec := kv.timestamp / 1e9
+	nsec := kv.timestamp - sec*1e9
```

Comment:
Habit. The above is faster, though obviously it doesn't matter for this code. I'll change it.